### PR TITLE
Task 1 - Helloworld

### PR DESCRIPTION
**Pull request by chlu0571**

### CODE REVIEW

**Comment by chlu0571** [comment]
I had created 3 main.nf (main/ main-1/ main-2) to perform the helloworld task. Can you check my scripts and tell me what to improve for the scripts? Thank you.

**Review by georgiesamaha — CHANGES_REQUESTED** [review]
A few things before I can test: 

1. Create a file in the main directory fastqc-nf called `.gitignore` and add the following to it. This will ensure you dont push all your nextflow logs and hidden files to the repo: 

```bash
.nextflow*
work/
```

2. Write some instructions in the User Guide section of the README.md file so I know how to execute the scripts. We can overwrite this section of the README.md for each new build you do. Can you write instructions using [markdown format](https://www.markdownguide.org/basic-syntax/), so include code inside code blocks.

File: README.md
Comment (georgiesamaha) [inline README.md:1]: Lets follow the general structure of this [workflow documentation guide](https://github.com/AustralianBioCommons/doc_guidelines/blob/master/documentation_templates/workflows.md). Don't have to contain all sections, just the necessary ones. See [here](https://github.com/Sydney-Informatics-Hub/Germline-StructuralV-nf) for an example.

File: README.md [hunk 1]
```diff
@@ -1,17 +1,22 @@
...
+## Run Scripts
+For main.nf
+```
+nextflow run main.nf
+```
 
-## User guide
+For main-1.nf
+```
+nextflow run -e.mode=ciao main-1.nf
```
Comment (georgiesamaha) [inline README.md:15]: What is the reasoning behind using `-e.mode`? I've never seen this before. Can you instead define a parameter at the top of `main.nf` called hello and apply it to the workflow with `nextflow run main.nf --hello ciao`

File: main.nf
Comment (georgiesamaha) [inline main.nf:1]: Can you please include the whole structure of this `main.nf` template with the help message, log info, comment lines, and execution summary? Have added additional comments below re: what to keep and what to adjust.

File: main.nf [hunk 2]
```diff
@@ -1,114 +1,16 @@
...
-
-=======================================================================================
-Workflow run parameters 
-=======================================================================================
-input       : ${params.input}
-outDir      : ${params.outDir}
-workDir     : ${workflow.workDir}
-=======================================================================================
-
-"""
```
Comment (georgiesamaha) [inline main.nf:45]: Keep this section, just adjust as required. For example: 

```
=======================================================================================
FastQC-nf
=======================================================================================
Created by Sydney Informatics Hub, University of Sydney  
Find documentation @ https://github.com/Sydney-Informatics-Hub/fastqc-nf
Cite this pipeline @ 
=======================================================================================
Workflow run parameters 
=======================================================================================
input       : ${params.hello}
workDir     : ${workflow.workDir}
=======================================================================================
```

File: main.nf [hunk 3]
```diff
@@ -1,114 +1,16 @@
...
-
-/// Help function
```
Comment (georgiesamaha) [inline main.nf:47]: Keep this comment line, same reason as above.

File: main.nf [hunk 4]
```diff
@@ -1,114 +1,16 @@
...
+process helloworld {
+  input: 
+    val x
+  output:
+    stdout
+  script:
+    """
+    echo '$x world!'
+    """
 }
```
Comment (georgiesamaha) [inline main.nf:13]: Keep this also, again adjust contents as required, for example: 

```
  Usage:  nextflow run main.nf --hello ciao
  Required Arguments:
  --hello Version of hello. If there are limited options here, specify what they are. 
```

File: main.nf [hunk 5]
```diff
@@ -1,114 +1,16 @@
...
-
-// Define workflow structure. Include some input/runtime tests here.
```
Comment (georgiesamaha) [inline main.nf:67]: Keep this comment line, same reason as above.

File: main.nf [hunk 6]
```diff
@@ -1,114 +1,16 @@
...
-// See https://www.nextflow.io/docs/latest/dsl2.html?highlight=workflow#workflow
 workflow {
-
-// Show help message if --help is run or (||) a required parameter (input) is not provided
```
Comment (georgiesamaha) [inline main.nf:71]: Keep this comment line, same reason as above.

File: main.nf [hunk 7]
```diff
@@ -1,114 +1,16 @@
...
-
-if ( params.help || params.input == false ){   
-// Invoke the help function above and exit
-	helpMessage()
-	exit 1
```
Comment (georgiesamaha) [inline main.nf:76]: This section will print your help message defined above. Keep it to print to the screen when you run `nextflow run main.nf --help`

File: main.nf [hunk 8]
```diff
@@ -1,114 +1,16 @@
...
-	// consider adding some extra contigencies here.
-	// could validate path of all input files in list?
-	// could validate indexes for reference exist?
-
-// If none of the above are a problem, then run the workflow
```
Comment (georgiesamaha) [inline main.nf:81]: Keep this comment line, same reason as above.

File: main.nf [hunk 9]
```diff
@@ -1,114 +1,16 @@
...
-Success     : ${workflow.success}
-workDir     : ${workflow.workDir}
-Exit status : ${workflow.exitStatus}
-outDir      : ${params.outDir}
-
-=======================================================================================
-  """
-println summary
-
-}
```
Comment (georgiesamaha) [inline main.nf:114]: Keep all of this to summarise workflow success. Remove outDir for this example.

File: main.nf [hunk 10]
```diff
@@ -1,114 +1,16 @@
...
-if ( params.help || params.input == false ){   
-// Invoke the help function above and exit
-	helpMessage()
-	exit 1
-	// consider adding some extra contigencies here.
-	// could validate path of all input files in list?
-	// could validate indexes for reference exist?
-
-// If none of the above are a problem, then run the workflow
-} else {
```
Comment (georgiesamaha) [inline main.nf:82]: Include this to run your pipeline if you do not use `--help`. You will need to include `Channel.of('Ciao', 'Hello', 'Hola') | helloworld | view` after it.

File: main.nf [hunk 11]
```diff
@@ -1,114 +1,16 @@
...
-	
-// Define channels
```
Comment (georgiesamaha) [inline main.nf:84]: Keep this comment line, same reason as above.

**Review by georgiesamaha — CHANGES_REQUESTED** [review]
Having your `nextflow.config` file inside the `fastqc-nf` directory will automatically apply it to your workflow execution. When I executed your second example command after executing the first, I got this message. Inside the `nextflow.config`, you're setting resource allocations, creating execution report, timeline, and trace files, and defining profiles. All files were saved to a directory called `results`. 

```
N E X T F L O W  ~  version 22.10.3
Launching `main-2.nf` [hungry_austin] DSL2 - revision: ba63f412ba
Trace file already exists: /home/ubuntu/fastqc-nf/results/runInfo/trace.txt -- enable the relevant `overwrite` option in your config file to overwrite existing files
```

Before we implement workflow configuration, might be worth deleting the `nextflow.config` from the repo and copying it across only when you are ready to apply it to your workflow.

**Pull request closed without merging**